Added security reporting guidelines. (#19201)
django/django: docs/internals/security.txt

team <https://www.djangoproject.com/foundation/teams/#security-team>`_.

Once you've submitted an issue via email, you should receive an acknowledgment

> from a member of the security team within 48 hours, and depending on the

Author: Yes, we have deliberately dropped "within 48 hours" but kept that will still receive the acknowledgement.
We can receive reports late on Friday, being followed up on Monday linking to "you should receive an acknowledgement within 48 hours".

> from a member of the security team within 48 hours, and depending on the

Author: Natalia and I do have working days so I think it is quite relevant still :+1: will push a tweak

analysis. Depending on the action to be taken, you may receive further followup

> emails. It can take several weeks before the security team has come to a

Author: We have received pushy replies before, even folks threatening to disclose publicly.
Do you have a wording suggestion?

> emails. It can take several weeks before the security team has come to a

Reviewer: > folks threatening to disclose publicly

Urgh :see_no_evil: 

Perhaps something along the lines of:

> It can take several weeks before the security team has come to a conclusion, depending on the :ref:`severity level <severity-levels>`. There is no need to chase the security team unless you discovery new, relevant information. All reports aim to be resolved within the industry-standard 90 days.

I'm not sure if there's a standard reply template, but including similar wording there might be a good idea, too.

> emails. It can take several weeks before the security team has come to a

Author: [Updated](https://code.djangoproject.com/wiki/SecurityTeam) :star:

team <https://www.djangoproject.com/foundation/teams/#security-team>`_.

Once you've submitted an issue via email, you should receive an acknowledgment

> from a member of the security team within 3 working days. After that, the
> security team will begin their analysis. Depending on the action to be taken,
> you may receive further followup emails. It can take several weeks before the
> security team has come to a conclusion. There is no need to chase the security
> team unless you discovery new, relevant information. All reports aim to be

Reviewer 2: ```suggestion
from a member of the security team within 3 working days. After that, the
security team will begin their analysis. Depending on the action to be taken,
you may receive followup emails. It can take several weeks before the security
team comes to a conclusion. There is no need to chase the security team
unless you discover new, relevant information. All reports aim to be
```

resolved within the industry-standard 90 days. Confirmed vulnerabilities with a

> :ref:`high severity level <severity-levels>` will be addressed promptly.

Reviewer: Thought: Everyone's favourite: bug bounties. Does Django have one? If not, this might be the time to mention it as a guideline.

I recently added [a similar note for Wagtail's security policy](https://docs.wagtail.org/en/stable/contributing/security.html#bug-bounties). 

(If this is out of scope, I'm happy to open a separate PR).

Author: We do, it's described here: https://hackerone.com/django?type=team
We talk about it when asked, not sure if we should add it as a section in case it encourages more "beg bounties"

Reviewer 3: > unless you discovery new

Typo: discover

up to roughly:

* `4k bytes for a URL`_.
* `8K bytes for a request header`_.

Other web servers, such as Nginx and Apache, have similar restrictions to
prevent excessive resource consumption.

Consequently, the Django security team will not consider reports that rely on
request headers or URLs exceeding 8K bytes, as such inputs are already
mitigated at the server level in production environments.

.. admonition:: :djadmin:`runserver` should never be used in production

    Django's built-in development server does not enforce these limits because
    it is not designed to be a production server.

.. _`4k bytes for a URL`: https://docs.gunicorn.org/en/stable/settings.html#limit-request-line
.. _`8k bytes for a request header`: https://docs.gunicorn.org/en/stable/settings.html#limit-request-field-size

The request body must be under 2.5MB
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :setting:`DATA_UPLOAD_MAX_MEMORY_SIZE` setting limits the default maximum
request body size to 2.5MB.

As this is enforced on all Django projects by default, a proof of concept must
not exceed 2.5MB in the request body to be considered valid.

> Code under test must feasibly exist in a Django project

Reviewer: Thought: Similar to the `runserver` note above, is it worth stating "production Django project" here? If a vulnerability only exists when other production-grade recommendations are ignored, would it be considered a legitimate issue? For example, anything requiring `DEBUG = True` (not that I can think of many cases).

Author: Updated, thank you

> If a vulnerability only exists when other production-grade recommendations are ignored, would it be considered a legitimate issue? 

No, in general the Django security team aims to mitigate vulnerabilities that could impact production deployed projects.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Django Template Language (DTL) is designed for building the content needed
to display web pages. In particular its text filters are meant for that kind of
usage.

For reference, the complete works of Shakespeare have about 3.5 million bytes
in plain-text ASCII encoding. Displaying such in a single request is beyond the
scope of almost all websites, and so outside the scope of the DTL too.

Text processing is expensive. Django makes no guarantee that DTL text filters
are never subject to degraded performance if passed deliberately crafted,
sufficiently large inputs. Under default configurations, Django makes it
difficult for sites to accidentally accept such payloads from untrusted
sources, but, if it is necessary to display large amounts of user provided
content, it’s important that basic security measures are taken.

User provided content should always be constrained to known maximum length. It
should be filtered to remove malicious content, and validated to match expected
formats. It should then be processed offline, if necessary, before being
displayed.

> Proof on concepts which use over 100K bytes of data to be processed by the DTL

Reviewer 4: Proof of concepts...

.. _our public Trac instance: https://code.djangoproject.com/query

> Reporting guidelines

Reviewer: Thought: Should these "guidelines" have a bit of text explaining that if someone is unsure whether an issue meets the guidelines, to err on the side of caution and submit "responsibly" anyway? In that it's better to fail safe. 

It also makes the "guidelines" feel slightly less like rules (which they're not).

Author: We have this paragraph within "How does Django evaluate a report"

> If you are unsure whether your finding meets these criteria, please still report
it :ref:`privately by emailing security@djangoproject.com
<reporting-security-issues>`. The security team will review your report and
recommend the correct course of action.

I think it's ok to keep that there.
I would be very pleasantly surprised is people have read these guidelines.
I expect that we link to points in these guidelines to explain why we have concluded that their report is not valid/they need to take these into account and send us an update

limits on request header and URL sizes. For example, by default Gunicorn allows
up to roughly:

* `4k bytes for a URL`_.
* `8K bytes for a request header`_.

Other web servers, such as Nginx and Apache, have similar restrictions to
prevent excessive resource consumption.

Consequently, the Django security team will not consider reports that rely on
request headers or URLs exceeding 8K bytes, as such inputs are already
mitigated at the server level in production environments.

.. admonition:: :djadmin:`runserver` should never be used in production

    Django's built-in development server does not enforce these limits because
    it is not designed to be a production server.

.. _`4k bytes for a URL`: https://docs.gunicorn.org/en/stable/settings.html#limit-request-line
.. _`8k bytes for a request header`: https://docs.gunicorn.org/en/stable/settings.html#limit-request-field-size

The request body must be under 2.5MB
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :setting:`DATA_UPLOAD_MAX_MEMORY_SIZE` setting limits the default maximum
request body size to 2.5MB.

As this is enforced on all production-grade Django projects by default, a proof

> of concept must not exceed 2.5MB in the request body to be considered valid.

Reviewer 3: It might be worth adding a note that issues resulting from large, but potentially reasonable setting values should be reported using the public tracker for hardening.

Code under test must feasibly exist in a Django project
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A proof of concept must demonstrate a realistic security issue - one that could

> affect a real-world Django application following standard development

Reviewer 2: Does "real-world" here mean something other than "production-grade"?

Reviewer 2: >Here I wanted to emphasize that the POC must be realistic in a web application.

How about:

```
The proof of concept must plausibly occur in a production-grade Django
application, reflecting real-world scenarios and following standard
development practices.
```

its public API. If a vulnerability depends on directly calling these internal
functions in an unsafe way, it will not be considered a valid security issue.

Content displayed by the Django Template Language must be under 100K bytes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Django Template Language (DTL) is designed for building the content needed
to display web pages. In particular its text filters are meant for that kind of
usage.

For reference, the complete works of Shakespeare have about 3.5 million bytes
in plain-text ASCII encoding. Displaying such in a single request is beyond the
scope of almost all websites, and so outside the scope of the DTL too.

Text processing is expensive. Django makes no guarantee that DTL text filters
are never subject to degraded performance if passed deliberately crafted,
sufficiently large inputs. Under default configurations, Django makes it
difficult for sites to accidentally accept such payloads from untrusted
sources, but, if it is necessary to display large amounts of user provided
content, it’s important that basic security measures are taken.

> User provided content should always be constrained to known maximum length. It

Reviewer 2: ```suggestion
User-provided content should always be constrained to known maximum length. It
```

                ["admin@example.com"],
            )
            return JsonResponse(status=200)
        return JsonResponse(form.errors, status=400)

Similarly, as Django's raw SQL constructs (such as :meth:`~.QuerySet.extra` and
:class:`.RawSQL` expression) provide developers with full control over the
query, they are insecure if user input is not properly handled. As explained in
our :ref:`security documentation <sql-injection-protection>`, it is the
developer's responsibility to safely process user input for these functions.

For instance, the following is **not considered valid** because ``query`` has
not been sanitized::

    from django.shortcuts import HttpResponse
    from .models import MyModel

    def my_proof_of_concept(request):
        query = request.GET.get("query", "")
        q = MyModel.objects.extra(select={"id": query})
        return HttpResponse(q.values())

Request headers and URLs must be under 8K bytes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To prevent denial-of-service (DoS) attacks, production-grade servers impose
limits on request header and URL sizes. For example, by default Gunicorn allows
up to roughly:

> * `4k bytes for a URL`_.
> * `8K bytes for a request header`_.

Reviewer 5: I'd remove dots
```suggestion
* `4k bytes for a URL`_
* `8K bytes for a request header`_
```

Other web servers, such as Nginx and Apache, have similar restrictions to
prevent excessive resource consumption.

Consequently, the Django security team will not consider reports that rely on
request headers or URLs exceeding 8K bytes, as such inputs are already
mitigated at the server level in production environments.

.. admonition:: :djadmin:`runserver` should never be used in production

    Django's built-in development server does not enforce these limits because
    it is not designed to be a production server.

.. _`4k bytes for a URL`: https://docs.gunicorn.org/en/stable/settings.html#limit-request-line
.. _`8k bytes for a request header`: https://docs.gunicorn.org/en/stable/settings.html#limit-request-field-size

> The request body must be under 2.5MB
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> The :setting:`DATA_UPLOAD_MAX_MEMORY_SIZE` setting limits the default maximum
> request body size to 2.5MB.
> 
> As this is enforced on all production-grade Django projects by default, a proof
> of concept must not exceed 2.5MB in the request body to be considered valid.

Reviewer 5: ```suggestion
The request body must be under 2.5 MB
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :setting:`DATA_UPLOAD_MAX_MEMORY_SIZE` setting limits the default maximum
request body size to 2.5 MB.

As this is enforced on all production-grade Django projects by default, a proof
of concept must not exceed 2.5 MB in the request body to be considered valid.
```

Issues resulting from large, but potentially reasonable setting values, should
be reported using the `public ticket tracker`_ for hardening.

.. _public ticket tracker: https://code.djangoproject.com/

Code under test must feasibly exist in a Django project
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A proof of concept must demonstrate a realistic security issue - one that could
affect a production-grade Django application following standard development
practices.

Django contains many private and undocumented functions that are not part of
its public API. If a vulnerability depends on directly calling these internal
functions in an unsafe way, it will not be considered a valid security issue.

> Content displayed by the Django Template Language must be under 100K bytes
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 5: ```suggestion
Content displayed by the Django Template Language must be under 100 KB
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
